### PR TITLE
mdns: MdnsConfig derives Clone and Debug.

### PR DESCRIPTION
**Pull request by dvc94ch**
Towards #2006 unblocks a new ipfs-embed release.

### CODE REVIEW

File: swarm/src/lib.rs [hunk 1]
```diff
@@ -412,6 +414,7 @@ where TBehaviour: NetworkBehaviour<ProtocolsHandler = THandler>,
     /// [`NetworkBehaviourAction::ReportObservedAddr`] or explicitly
     /// through this method.
     pub fn add_external_address(&mut self, a: Multiaddr, s: AddressScore) -> AddAddressResult {
+        self.behaviour.inject_new_external_addr(&a);
         self.external_addrs.add(a, s)
```
Comment (romanb) [inline swarm/src/lib.rs:418]:
```suggestion
        let result = self.external_addrs.add(a, s);
        if let AddAddressResult::Inserted = &result {
            self.behaviour.inject_new_external_addr(&a);
        }
        result
```

File: swarm/src/lib.rs [hunk 2]
```diff
@@ -422,6 +425,7 @@ where TBehaviour: NetworkBehaviour<ProtocolsHandler = THandler>,
     /// Returns `true` if the address existed and was removed, `false`
     /// otherwise.
     pub fn remove_external_address(&mut self, addr: &Multiaddr) -> bool {
+        self.behaviour.inject_expired_external_addr(addr);
         self.external_addrs.remove(addr)
```
Comment (romanb) [inline swarm/src/lib.rs:429]:
```suggestion
        if self.external_addrs.remove(addr) {
            self.behaviour.inject_expired_external_addr(addr);
        }
```
However, note also that the most common way in which external addresses expire is as a result of new addresses being added. See https://github.com/libp2p/rust-libp2p/blob/master/swarm/src/registry.rs#L186-L196. So, to be able to properly invoke this callback whenever an external address expires, I think you would need to incorporate the expired/removed addresses within a call to `add()` in the `AddAddressResult`, so that the `Swarm` can then invoke the callback.

**Comment by romanb** [comment]
The additions generally look good to me. We just need to make sure that the new callbacks for external addresses are called only (and always) when appropriate.

File: swarm/src/behaviour.rs [hunk 3]
```diff
@@ -147,17 +147,17 @@ pub trait NetworkBehaviour: Send + 'static {
...
+    /// Indicates to the behaviour that a new listener was created.
+    fn inject_new_listener(&mut self, _id: ListenerId) {
+    }
+
     /// Indicates to the behaviour that we have started listening on a new multiaddr.
-    fn inject_new_listen_addr(&mut self, _addr: &Multiaddr) {
+    fn inject_new_listen_addr(&mut self, _id: ListenerId, _addr: &Multiaddr) {
     }
 
     /// Indicates to the behaviour that a new multiaddr we were listening on has expired,
```
Comment (mxinden) [inline swarm/src/behaviour.rs:158]:
```suggestion
    /// Indicates to the behaviour that a multiaddr we were listening on has expired,
```

File: protocols/mdns/src/behaviour.rs [hunk 4]
```diff
@@ -49,6 +49,8 @@ lazy_static! {
         SocketAddr::from((Ipv4Addr::new(224, 0, 0, 251), 5353));
 }
 
+/// Configuration for mDNS.
+#[derive(Clone, Debug)]
```
Comment (mxinden) [inline protocols/mdns/src/behaviour.rs:53]: Could you extract this change to a separate pull request? Making only a single cohesive change per pull request makes maintaining this project a lot easier, especially around version and changelog management.

Comment (dvc94ch) [reply protocols/mdns/src/behaviour.rs:53]: I see you renamed the PR, I moved the api changes, as this one has the branch name `mdns-config`
